### DISCUSSION TITLE
Combine Tonic with Actix-web

### DISCUSSION THREAD

**Discussion by Leanerz** [hyperium/tonic#704] [Q&A]
Hi

Is there a way to combine properly Tonic and Actix-web?

The goal is to get a GRPC client as pool and pass it to actix-web to use it, or even make a shared GRPC client without opening each time connection for each requests of users.

Any idea please to follow?

Thank you !

**Comment by davidpdrsn** [comment]
Hm not quite sure what you're asking. It should be possible to create a tonic client, the way you always do it, and somehow pass it to your actix-web app. Maybe using [`App::data`](https://docs.rs/actix-web/3.3.2/actix_web/struct.App.html#method.data)? I'm not experienced actix-web myself so I really don't know.

@Leanerz does that help? Otherwise, could you try and be more specific? Perhaps share some code?

**Comment by gold-wallet** [comment]
> Hm not quite sure what you're asking. It should be possible to create a tonic client, the way you always do it, and somehow pass it to your actix-web app. Maybe using [`App::data`](https://docs.rs/actix-web/3.3.2/actix_web/struct.App.html#method.data)? I'm not experienced actix-web myself so I really don't know.
> 
> @Leanerz does that help? Otherwise, could you try and be more specific? Perhaps share some code?

Does tonic support pooling? He asks to pass a pool instead of simple client. Interested as well.

**Comment by davidpdrsn** [comment]
A tonic client is backed by a [`tonic::transport::Channel`](https://docs.rs/tonic/0.4.3/tonic/transport/channel/struct.Channel.html) which by default establishes a single h2 connection to the host. Because of h2 you can send multiple requests in parallel (thanks to multiplexing) and clients can be shared by cloning them. Cloning is cheap and the underlying connection is reused. But it doesn't maintain a pool of multiple connections to the same host similar to what [`hyper::Client`](https://docs.rs/hyper/0.14.9/hyper/client/struct.Client.html) does.

You can however load balance across multiple endpoints using [`Channel::balance_list`](https://docs.rs/tonic/0.4.3/tonic/transport/channel/struct.Channel.html#method.balance_list).

**Reply by Leanerz** [reply]
Hi

I passed a GRPC client to ACTIX-WEB this way:

```
    //GRPC
    let endpoints = ["http://[::1]:50051", "http://[::1]:50052"].iter().map(|a| Channel::from_static(a));

    let channel = Channel::balance_list(endpoints);

    let grpc_client = grpcClient::new(channel);

   HttpServer::new(move || {
        App::new()
            .data(grpc_client.clone())
   ...
```

But there is no way to get it from inside a router:
```

pub struct UserSTRUCT {
    pub grpc_client: &'static web::Data<tonic::transport::Channel>,
}

impl TUser for UserSTRUCT {
    fn RemoveUser(&self, user: u64) -> bool {
        

		let j = tokio::spawn(async move {
			let request = tonic::Request::new(self.grpc_client.new_req {
           		message: "hello".into(),
        	});
        

        	let response = self.grpc_client.unary_echo(request).await;

        	println!("RESPONSE={:?}", response);
        });

        true
    }
}
```

But grpc_client not recognized !

Any idea about this? Thank you again !

**Reply by davidpdrsn** [reply]
I've asked on discord of anyone is able to help. I know pretty much nothing about actix.

**Reply by davidpdrsn** [reply]
Asking in one of places listed here might also help https://actix.rs/community/

**Reply by Leanerz** [reply]
Thank you ! Really appreciated !

**Reply by byblakeorriver** [reply]
It looks like you are missing the grpcClient in your struct. 

It should look something like this:
```
pub struct UserSTRUCT {
    pub grpc_client: grpcClient<tonic::transport::Channel>,
}
```

That is why it is complaining that grpc_client is not recognized, I think.

However, I think you can get away without using a struct at all. If you just put the web::Data in the function arguments like this

```
fn RemoveUser(&self, user: u64, web::Data<grpcClient<tonic::transport::Channel>>) -> bool
```

That is if you are using something like this in you App builder:
```
App::new()
      .data(grpc_client.clone())
      .route(
        "/api/user/remove",
        web::post().to(RemoveUser),
      )
```

Just adding the client via `.data()` puts the client in Arc that can be used safely across threads.
